utils/tf_solver.py

```diff
@@ -60,7 +101,31 @@ def get_next(stop_val):
 
     def run(self, callback):
         self.model = self.model_init_fn()
-        self.optimizer = self.optimizer_klass(**self.optimizer_kwargs)
+        self.optimizer = self.optimizer_klass(
+            # this scaling is needed as in TF the weight decay is
+            # not multiplied by the learning rate
+            weight_decay=self.decoupled_weight_decay*self.lr,
+            **self.optimizer_kwargs,
+        )
+        if self.coupled_weight_decay:
+            # this is equivalent to adding L2 regularization to all
+            # the weights and biases of the model (even if adding
+            # weight decay to the biases is not recommended), of a factor
+            # halved
+            l2_reg_factor = self.coupled_weight_decay / 2
+            # taken from
+            # https://sthalles.github.io/keras-regularizer/
+            regularizer = tf.keras.regularizers.l2(l2_reg_factor)
+            target_regularizers = [
+                'kernel_regularizer',
+                'bias_regularizer',
+                'beta_regularizer',
+                'gamma_regularizer',
+            ]
+            for layer in self.model.layers:
+                for attr in target_regularizers:
+                    if hasattr(layer, attr):
+                        setattr(layer, attr, regularizer)
```

Comment:
This seems very custom, why not use the `weight_decay` from the optimizer?

Reply:
There is no `weight_decay` argument in `Adam`, so we need to use this API to get weight decay.